txindex/filter.go

```diff
@@ -1,22 +1,117 @@
 package txindex
 
-import "github.com/raedahgroup/dcrlibwallet/txhelper"
+import (
+	"github.com/asdine/storm/q"
+	"github.com/raedahgroup/dcrlibwallet/txhelper"
+)
 
 type ReadFilter struct {
-	typeFilter      []string
-	directionFilter []txhelper.TransactionDirection
+	matcher q.Matcher
 }
 
 func Filter() *ReadFilter {
 	return &ReadFilter{}
 }
 
-func (filter *ReadFilter) WithTxTypes(txTypes ...string) *ReadFilter {
-	filter.typeFilter = txTypes
+func (filter *ReadFilter) AndWithTxTypes(txTypes ...string) *ReadFilter {
+	var filterItems []interface{}
+	for _, txDirection := range txTypes {
+		filterItems = append(filterItems, txDirection)
+	}
+
+	filter.matcher = filter.createMatcher("Direction", filterItems, q.And)
+	return filter
+}
+
+func (filter *ReadFilter) AndForDirections(txDirections ...txhelper.TransactionDirection) *ReadFilter {
+	var filterItems []interface{}
+	for _, txDirection := range txDirections {
+		filterItems = append(filterItems, txDirection)
+	}
+
+	filter.matcher = filter.createMatcher("Direction", filterItems, q.And)
+
+	return filter
+}
+
+func (filter *ReadFilter) OrWithTxTypes(txTypes ...string) *ReadFilter {
+	var filterItems []interface{}
+	for _, txDirection := range txTypes {
+		filterItems = append(filterItems, txDirection)
+	}
+
+	filter.matcher = filter.createMatcher("Direction", filterItems, q.Or)
+	return filter
+}
+
+func (filter *ReadFilter) OrForDirections(txDirections ...txhelper.TransactionDirection) *ReadFilter {
+	var filterItems []interface{}
+	for _, txDirection := range txDirections {
+		filterItems = append(filterItems, txDirection)
+	}
+
+	filter.matcher = filter.createMatcher("Direction", filterItems, q.Or)
+
 	return filter
 }
 
-func (filter *ReadFilter) ForDirections(txDirections ...txhelper.TransactionDirection) *ReadFilter {
-	filter.directionFilter = txDirections
+func (filter *ReadFilter) createMatcher(fieldName string, items []interface{}, joinFunc func(matchers ...q.Matcher) q.Matcher) q.Matcher {
```

Comment:
- Rename `items` to `values`, `fieldValues` or `compareValues`
- Rename `joinFunc` to `joinMatchers` or `combineMatchers`